File: files/400-499/443/CH-443.R
library(tidyverse)
library(readxl)

path <- "400-499/443/CH-443 Number Puzzles - Disarium Number.xlsx"
input <- read_excel(path, range = "B3:B10")
test <- read_excel(path, range = "D3:D10")

is_disarium <- function(n) {
  digits <- as.integer(strsplit(as.character(n), "")[[1]])
  n == sum(digits^seq_along(digits))
}

canditates <- 0:1e6 %>%
  keep(is_disarium)
result <- input %>%
  crossing(canditates) %>%
  summarize(
    result = case_when(
      any(canditates == Number) ~ max(canditates[canditates < Number]),
      .default = canditates[which.min(abs(canditates - Number))]
    ),
    .by = Number
  )
# different results

File: files/400-499/443/CH-443.py
import pandas as pd

path = "400-499/443/CH-443 Number Puzzles - Disarium Number.xlsx"
input = pd.read_excel(path, usecols="B", nrows=7, skiprows=2)
test = pd.read_excel(path, usecols="D", nrows=7, skiprows=2)


def is_disarium(n):
    digits = map(int, str(n))
    return n == sum(digit**position for position, digit in enumerate(digits, 1))


candidates = [n for n in range(1_000_001) if is_disarium(n)]
result = (
    input.assign(key=1)
    .merge(pd.DataFrame({"candidate": candidates, "key": 1}), on="key")
    .drop(columns="key")
    .groupby("Number", as_index=True)
    .apply(
        lambda df: pd.Series(
            {
                "result": (
                    df.loc[df["candidate"] < df.name, "candidate"].max()
                    if (df["candidate"] == df.name).any()
                    else df.loc[(df["candidate"] - df.name).abs().idxmin(), "candidate"]
                )
            }
        )
    )
    .reset_index()
)
